### PR TITLE
#19147 : The evaluation order in Vanity URLs is still being used by s…

### PR DESCRIPTION
**Pull request by jcastro-dotcms**
…everal customers. The new version of the Vanity URL API was not setting the expected value for that. The API is now returning the Vanities in ascending order so that users can specify the appropriate way in which Vanities should match an incoming URL.

### CODE REVIEW

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java [hunk 1]
```diff
@@ -125,22 +126,18 @@ public void populateAllVanityURLsCache() throws DotDataException {
           vanityUrls.stream().map(vanity -> vanity.get("live_inode").toString()).collect(Collectors.toList());
       final List<Contentlet> contentlets = this.contentletAPI.findContentlets(vanityUrlInodes);
 
-      return contentlets.stream().map(contentlet -> new CachedVanityUrl(this.fromContentlet(contentlet))).collect(Collectors.toList());
+      return contentlets.stream().map(contentlet -> new CachedVanityUrl(this.fromContentlet(contentlet))).sorted().collect(Collectors.toList());
 
     } catch (final Exception e) {
       Logger.error(this,
-          String.format("An error occurred when retrieving Vanity URLs: siteId=[%s], " + "languageId=[%s], includeSystemHost=[%s]",
+          String.format("An error occurred when retrieving Vanity URLs: siteId=[%s], languageId=[%s]",
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java:133]: We should always append `e.getMessage()` to our message so the real root cause does not get eaten higher up.

**Review by wezell — CHANGES_REQUESTED** [review]
Looks great, one minor thing.

**Comment by Xander-Steinmann** [comment]
Does this code change mean that before the fix the ordering never worked?

**Comment by wezell** [comment]
Apparently, ordering worked until the Vanity URL re-write that happened in 5.3.1, which was done to fix other problems with the vanities.

File: dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java [hunk 2]
```diff
@@ -73,6 +72,65 @@ public static void prepare() throws Exception {
...
+                        .stream()
+                        .filter(v -> v.forwardTo.equals(baseUri))
+                        .collect(Collectors.toList());
+
+        // testing we have all the vanities we just saved
+        assert(cachedVanities.size() == numberOfTestVanities);
+
+        
+        // testing they have different order
+        assert(cachedVanities.get(0).order < cachedVanities.get(numberOfTestVanities-1).order);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java:112]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6064444)

File: dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java [hunk 3]
```diff
@@ -73,6 +72,65 @@ public static void prepare() throws Exception {
...
+     * tests that we correctly order the cached vanites.  This ordering is done
+     * by the findInDb method, which should return a hosts' vanities in order asc
+     * 
+     * It also tests that the vanityUrlAPI.resolveVanityUrl returns
+     * the first matching vanity in the list
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void testing_vanityurl_ordering() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java:85]: Codacy found an issue: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6064444)

File: dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java [hunk 4]
```diff
@@ -73,6 +72,65 @@ public static void prepare() throws Exception {
...
+
+
+        // this list of CachedVanityUrl is ordered based on the order of the vanities - asc
+        List<CachedVanityUrl> cachedVanities = vanityUrlAPI.findInDb(defaultHost, defaultLanguage)
+                        .stream()
+                        .filter(v -> v.forwardTo.equals(baseUri))
+                        .collect(Collectors.toList());
+
+        // testing we have all the vanities we just saved
+        assert(cachedVanities.size() == numberOfTestVanities);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java:108]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6064444)

File: dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java [hunk 5]
```diff
@@ -73,6 +72,65 @@ public static void prepare() throws Exception {
...
+        // testing they have different order
+        assert(cachedVanities.get(0).order < cachedVanities.get(numberOfTestVanities-1).order);
+        
+
+
+        CachedVanityUrl previousCachedVanity = cachedVanities.get(0);
+        
+        // the vanities are in order, from lowest to highest
+        for (final CachedVanityUrl cached : cachedVanities) {
+            assert(cached.order >= previousCachedVanity.order);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/vanityurl/business/VanityUrlAPITest.java:120]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6064444)

**Review by freddyucv — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

**Pull request merged**